sympy/matrices/expressions/vandermonde.py

```diff
@@ -0,0 +1,141 @@
+             \begin{bmatrix}
+             1 & x_1 & x_1^2 & \hdots % x_1^{n-1} \\
+             1 & x_2 & x_2^2 & \hdots % x_2^{n-1} \\
+            \vdots & \vdots & \vdots & \ddots & \vdots \\
+             1 & x_n & x_n^2 & \hdots % x_n^{n-1} \\
+             \end{bmatrix}
+
+        Parameters
+        ==========
+
+        points : iterable
+            The points to generate the VandermondeMatrix for.
+
+        n : integer, iterable or None
+            Powers to use. For an integer, the powers range from 0 to n-1.
+            For an iterable, use those powers. If ``None`` (default), the
+            powers range from 0 to ``len(points)``-1.
+
+        Examples
+        ========
+
+        >>> from sympy.matrices.expressions.vandermonde import VandermondeMatrix
+        >>> v = VandermondeMatrix([0, 1, -1])
+        >>> v.as_explicit()
+        Matrix([
+        [1,  0, 0],
+        [1,  1, 1],
+        [1, -1, 1]])
+
+        >>> from sympy import symbols
+        >>> x = symbols('x(1:4)')
+        >>> VandermondeMatrix(x, range(4)).as_explicit()
+        Matrix([
+        [1, x1, x1**2, x1**3],
+        [1, x2, x2**2, x2**3],
+        [1, x3, x3**2, x3**3]])
+
+        Sometimes, a Vandermonde matrix is defined without the power of zero
+        column:
+
+        >>> VandermondeMatrix(x, range(1,4)).as_explicit()
+        Matrix([
+        [x1, x1**2, x1**3],
+        [x2, x2**2, x2**3],
+        [x3, x3**2, x3**3]])
+
+        As it is possible to specify an iterable for the powers, generalized
+        versions can be rapildy created:
+
+        >>> VandermondeMatrix([-1, 2, 3], [1, 3, 6]).as_explicit()
+        Matrix([
+        [-1, -1,   1],
+        [ 2,  8,  64],
+        [ 3, 27, 729]])
+
+        It is also possible to use a symbolic column count:
+        >>> n = symbols('n')
+        >>> VandermondeMatrix([-1, 2, 3], n)
+        VandermondeMatrix([-1, 2, 3], n)
+
+        or symbolic powers
+
+        >>> VandermondeMatrix(x, [0, 1, n]).as_explicit()
+        Matrix([
+        [1, x1, x1**n],
+        [1, x2, x2**n],
+        [1, x3, x3**n]])
+
+        References
+        ==========
+
+        .. [1] https://en.wikipedia.org/wiki/Vandermonde_matrix
+
+    """
+    def __new__(cls, points, n=None):
+        if not iterable(points):
+            points = [_sympify(points)]
+        rows = len(points)
+        cls.rowvals = points
+        if n:
+            if iterable(n):
+                cols = len(n)
+                cls.colpows = n
+            else:
+                n = _sympify(n)
+                cols = n
+                if n.is_Integer:
+                    cls.colpows = range(n)
+                else:
+                    cls.colpows = Range(n)
+        else:
+            cols = rows
+            cls.colpows = range(cols)
+        cls._check_dim(rows)
+        cls._check_dim(cols)
+
+        cls._rows = _sympify(rows)
+        cls._cols = _sympify(cols)
+        obj = super().__new__(cls, rows, cols)
```

Comment:
In the sympy class designs, you should store everything as similar as the signature of object constructor 
`__new__(cls, points, n=None) -> super().__new__(cls, points, n)`.
And it's not a good idea to store as arbitrary attributes as `cls.rowvals`, `cls.colpows` and rely on this, but you should derive every information from `points, n`.

Reply:
I cannot pass the list of points to `MatExpr` as `_sympify` will complain. Most other parts are done. (I still believe it can make sense to store `cls.colpows` to avoid having to regenerate them in `_entry` on every call.)

Reply:
Can you convert the list of points to tuple?

Reply:
Ahh, thanks!